tests/test_browser.py

```diff
@@ -81,35 +81,65 @@ def test__request(httpbin):
         "Content-Type"]
 
 
-def test__request_file(httpbin):
-    form_html = """
-    <form method="post" action="{}/post">
-      <input name="pic" type="file" />
+valid_enctypes_file_submit = {"multipart/form-data": True,
+                              "application/x-www-form-urlencoded": False
+                              }
+
+default_enctype = "application/x-www-form-urlencoded"
+
+
+@pytest.mark.parametrize("try_submit_file", [
+  pytest.param(True), pytest.param(False)
+])
+@pytest.mark.parametrize("enctype", [
+  pytest.param("multipart/form-data"),
+  pytest.param("application/x-www-form-urlencoded"),
+  pytest.param("Invalid encytpe")
+])
+def test_enctype_and_file_submit(httpbin, enctype, try_submit_file):
+    # test if enctype is respected when specified
+    # and if files are processed correctly
+    form_html = ("""
+    <form method="post" action="{}/post" enctype="{}">
+      <input name="in" value="test" />
+      <input name="fileinput" type="file" />
     </form>
-    """.format(httpbin.url)
+    """).format(httpbin.url, enctype)
     form = BeautifulSoup(form_html, "lxml").form
 
-    # create a temporary file for testing file upload
-    pic_path = tempfile.mkstemp()[1]
-    with open(pic_path, "w") as f:
-        f.write(":-)")
+    if try_submit_file:
+        file_path = tempfile.mkstemp()[1]
+        with open(file_path, "w") as f:
+            f.write(":-)")
 
-    form.find("input", {"name": "pic"})["value"] = pic_path
+        form.find("input", {"name": "fileinput"})["value"] = file_path
 
     browser = mechanicalsoup.Browser()
     response = browser._request(form)
 
-    # Check that only "files" includes a "pic" keyword in the response
-    found = False
-    for key, value in response.json().items():
-        if key == "files":
-            assert value["pic"] == ":-)"
-            found = True
-        else:
-            assert (value is None) or ("pic" not in value)
-
-    assert found
-    assert "multipart/form-data" in response.request.headers["Content-Type"]
+    resp = response.json()
+    assert resp["form"]["in"] == "test"
+
+    if enctype not in valid_enctypes_file_submit.keys():
+        enctype = default_enctype
+
+    assert enctype in response.request.headers["Content-Type"]
+
+    # response must contain files if and only if the enctype support
+    # sending files and if we submitted files
+    if (valid_enctypes_file_submit[enctype] and try_submit_file):
+        # test if only "files" includes "fileinput" keyword
+        for key, value in resp.items():
+            if key == "files":
+                assert value["fileinput"] == ":-)"
+                found = True
+            else:
+                assert (value is None) or ("fileinput" not in value)
+
+        assert found
+
```

Comment:
Useless newline.

Reply:
I still find this newline useless and disturbing (I'd expect the `if` to be over when I see this newline, and I'm surprised by the `else:` right below).